### PR TITLE
add custom_jvp for jax.nn.softplus/log_sigmoid

### PR DESCRIPTION
**Pull request by DanisNone**

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi - can you say more about what problem this solves? Have you run into any situations where the current gradient fails?

**Comment by DanisNone** [comment]
@jakevdp Sorry, I should have added a description earlier.


The gradient works perfectly fine, but since it's computed automatically, it involves quite a lot of operations. Adding custom_jvp allows collapsing all of that into just a few operations.

I'm not sure how correct this is, but if you look at the output of jax.make_jaxpr, you can see that the version with custom_jvp requires far fewer instructions:

```python
import jax
import jax.numpy as jnp

@jax.jit
def f1(x):
    return jnp.logaddexp(x, 0)

@jax.custom_jvp
@jax.jit
def f2(x):
    return jnp.logaddexp(x, 0)
f2.defjvps(lambda g, ans, x: g * jax.lax.logistic(x))

grad_f1 = jax.grad(f1)
jaxpr1 = jax.make_jaxpr(grad_f1)(3.0)

grad_f2 = jax.grad(f2)
jaxpr2 = jax.make_jaxpr(grad_f2)(3.0)

print(jaxpr1)
print("-"*50)
print(jaxpr2)
```

**Comment by jakevdp** [comment]
Thanks - I think this is working as intended. In general we add custom derivative rules only in cases where it's required to get correct outputs. Regarding the many operations: we can generally count on the compiler to optimize the runtime.

You can go ahead and close this PR.

**Comment by DanisNone** [comment]
@jakevdp  Is there a way to see the result of compiler optimization?

**Comment by jakevdp** [comment]
It's described here: https://docs.jax.dev/en/latest/aot.html
